earthfile2llb/withdockerrunreg.go

```diff
@@ -0,0 +1,272 @@
+package earthfile2llb
+
+import (
+	"context"
+
+	"github.com/earthly/earthly/domain"
+	"github.com/earthly/earthly/states"
+	"github.com/earthly/earthly/util/llbutil/pllb"
+	"github.com/earthly/earthly/util/platutil"
+	"github.com/earthly/earthly/util/syncutil/semutil"
+	"github.com/moby/buildkit/client/llb"
+	"github.com/pkg/errors"
+)
+
+type withDockerRunRegistry struct {
+	*withDockerRunBase
+	c *Converter
+
+	enableParallel bool
+	sem            semutil.Semaphore
```

Comment:
It doesn't look like this semaphore is ever initialized. This should be very similar to the tar version:

```go
	// This semaphore ensures that there is at least one thread allowed to progress,
	// even if parallelism is completely starved.
	w.sem = semutil.NewMultiSem(w.c.opt.Parallelism, semutil.NewWeighted(1))
```

The reason this needs to have at least one thread is that all the parallelism assumes that the main thread can always make progress and never needs to wait for a secondary thread itself. In the case of the use of `BuildAsync` here, however, we are breaking that assumption - we are waiting for all the promises to finish before progressing work on the main thread. So to avoid a dead-lock, we ensure that at least one secondary thread (started via `BuildAsync`) can always make progress.

Reply:
Thanks. I must have missed that. I moved it to both constructors.